docs/rfc/rfc-010-p2p-light-client.rst

```diff
@@ -0,0 +1,127 @@
+==================================
+RFC 010: Peer to Peer Light Client
+==================================
+
+Changelog
+---------
+
+- 2022-01-21: Initial draft (@tychoish)
+
+Abstract
+--------
+
+The dependency on access to the RPC system makes running or using the light
+client more complicated than it should be, because it is difficult to get
+access to the RPC endpoints on a running node. There is no deep dependency for
+the light client on the RPC system, and there is a persistent notion that
+"make a p2p light client" is a solution to this operational limitation. This
+document explores the implications and requirements of implementing a
+p2p-based light client, as well as the possibilities afforded by this
+implementation.
+
+Background
+----------
+
+High Level Design
+~~~~~~~~~~~~~~~~~
+
+From a high level, the light client P2P implementation, is relatively straight
+forward, and the P2P-backed statesync implementation that took place during
+the 0.35 cycle. The light client only really needs to be able to request (and
+receive) a `LightBlock` at a given height. To support this, a new Reactor
+would run on every full node and validator which would be able to service
+these requests. The workload would be entirely request-response, and the
+implementation of the reactor would likely be very straight forward, and the
+implementation of the provider is similarly relatively simple. 
+
+The complexity of the project focuses around peer discovery, handling when
+peers disconnect from the light clients, and how to change the current P2P
+code to appropriately handle specialized nodes.
+
+I believe it's safe to assume that much of the current functionality of the
+current ``light`` mode would *not* need to be maintained: there is no need to
+proxy the RPC endpoints over the P2P layer and there may be no need to run a
+node/process for the p2p light client (e.g. all use of this will be as a
+client.) 
+
+The ability to run light clients using the RPC system will continue to be
+maintained.
+
+LibP2P
+~~~~~~
+
+While some aspects of the P2P light client implementation are orthogonal to
+LibP2P project, it's useful to think about the ways that these efforts may
+combine or interact.
+
+We expect to be able to leverage libp2p tools to provide some kind of service
+discovery for tendermint-based networks. This means that it will be possible
+for the p2p stack to easily identify specialized nodes, (e.g. light clients)
+thus obviating many of the design challenges with providing this feature in
+the context of the current stack.
+
+Similarly, libp2p makes it possible for project to be able back their non-Go
+light clients, without the major task of first implementing tendermint's p2p
+connection handling. We should identify if there exist users (e.g. the go IBC
+relayer, it's maintainers, and operators) who would be able to take advantage
```

Comment:
I talked to @jackzampolin about this feature a few times and they would love to incorporate this. The only missing part is a way to submit txs and query through the p2p based light client. 

For submitting txs this could be done on the mempool connection, but query would need to be something new.

Reply:
There aren't separate or per-reactor connections in the p2p layer so we'd need to build something out for this, and while we could add a "submit transaction over the p2p layer" (and perhaps should, I think this is a reasonable feature). 

I'd also be super curious to know what the number of deployments of the go relayer are, and getting much more information about that roadmap.